### PR TITLE
Core: Add DateFormatter interface for java time parsing

### PR DESCRIPTION
**Pull request by spinscale**
The existing approach used java time date formatters when a format based 
string like `date_time||epoch_millis` was used, instead of the custom 
code written in the epoch date formatter.

In order to properly solve this, a new interface called
`DateFormatter` has been added, which now can be implemented for custom
formatters. Currently there are two implementations, one using java time
and one doing the epoch_millis formatter, which simply parses a number
and then converts it to a date in UTC timezone.

The DateFormatter interface now also has a method to retrieve the name
of the formatter, which is needed for mapping changes anyway.

The existing `CompoundDateTimeFormatter` class has been removed, the
name was not really nice anyway.

One more minor change is the fact, that the new java time using
FormatDateFormatter does not try to parse the date with its printer
implementation first (which might be a strict one and fail), but a
printer can now be specified in addition. This saves one potential
failure/exception when parsing less strict dates.

If only a printer is specified, the printer will also be used as a
parser.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/time/DateFormatter.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+     */
+    DateFormatter withZone(ZoneId zoneId);
+
+    /**
+     * Print the supplied java time accessor in a string based representation according to this formatter
+     *
+     * @param accessor The temporal accessor used to format
+     * @return         The string result for the formatting
+     */
+    String print(TemporalAccessor accessor);
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/time/DateFormatter.java]: Can we stick with the terminology used in java time? I think this should be called `format`.

File: server/src/main/java/org/elasticsearch/common/time/DateFormatter.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    /**
+     * A name based format for this formatter. Can be one of the registered formatters like <code>epoch_millis</code> or
+     * a configured format like <code>HH:mm:ss</code>
+     *
+     * @return The name of this formatter
+     */
+    String format();
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/time/DateFormatter.java]: Can we call this `pattern`? Java doesn't have any concept of these named patterns, but that is essentially what it is.

File: server/src/main/java/org/elasticsearch/common/time/EpochMillisDateFormatter.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeParseException;
+import java.time.temporal.TemporalAccessor;
+import java.time.temporal.TemporalField;
+import java.util.Map;
+
+class EpochMillisDateFormatter implements DateFormatter {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/time/EpochMillisDateFormatter.java]: At least a single line javadoc please?

File: server/src/main/java/org/elasticsearch/common/time/FormatDateFormatter.java [hunk 4]
```diff
@@ -27,33 +28,28 @@
...
 import java.util.Locale;
 import java.util.Map;
-import java.util.function.Consumer;
 
-/**
- * wrapper class around java.time.DateTimeFormatter that supports multiple formats for easier parsing,
- * and one specific format for printing
- */
-public class CompoundDateTimeFormatter {
+class FormatDateFormatter implements DateFormatter {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/time/FormatDateFormatter.java]: Maybe just JavaDateFormatter since it uses the java time classes?

**Review by rjernst — COMMENTED** [review]
Looks pretty good. I have a few suggestions.

**Comment by spinscale** [comment]
took care of all suggestions. thanks for looking!

**Review by rjernst — APPROVED** [review]
LGTM!

**Pull request merged**